lib/exec.js

```diff
@@ -45,28 +48,46 @@ async function exec (cmd, args = [], opts = {}) {
         reject(new Error(`Standard input '${err.syscall}' error: ${err.stack}`));
       });
     }
+    let stdoutBufferSize = 0;
     if (proc.stdout) {
       proc.stdout.on('error', (err) => {
         reject(new Error(`Standard output '${err.syscall}' error: ${err.stack}`));
       });
       // keep track of stdout if we have not said not to
-      if (!opts.ignoreOutput) {
+      if (opts.ignoreOutput) {
+        // https://github.com/nodejs/node/issues/4236
+        proc.stdout.on('data', () => {});
+      } else {
         proc.stdout.on('data', (data) => {
           stdoutArr.push(data);
+          stdoutBufferSize += _.size(data);
+          while (stdoutArr.length > 1 && stdoutBufferSize >= opts.maxStdoutBufferSize) {
+            stdoutBufferSize -= _.size(_.first(stdoutArr));
+            stdoutArr.shift();
+          }
           if (opts.logger && _.isFunction(opts.logger.debug)) {
             opts.logger.debug(data);
           }
         });
       }
     }
+    let stderrBufferSize = 0;
```

Comment:
there's starting to be an awful lot of code which is duplicated for both stdout and stderr in this section. does it make sense to refactor for code reuse?

Reply:
I'll try to do that